core/src/test/java/org/apache/iceberg/TestRowLineageMetadata.java

```diff
@@ -0,0 +1,328 @@
+      assertThat(TableMetadata.buildFromEmpty(formatVersion).enableRowLineage()).isNotNull();
+    } else {
+      assertThatThrownBy(() -> TableMetadata.buildFromEmpty(formatVersion).enableRowLineage())
+          .isInstanceOf(IllegalArgumentException.class)
+          .hasMessageContaining("Cannot use row lineage");
+    }
+  }
+
+  @TestTemplate
+  public void testSnapshotAddition() {
+    assumeThat(formatVersion).isGreaterThanOrEqualTo(TableMetadata.MIN_FORMAT_VERSION_ROW_LINEAGE);
+
+    long newRows = 30L;
+
+    TableMetadata base = baseMetadata();
+
+    Snapshot addRows =
+        new BaseSnapshot(
+            0, 1, null, 0, DataOperations.APPEND, null, 1, "foo", base.nextRowId(), newRows);
+
+    TableMetadata firstAddition = TableMetadata.buildFrom(base).addSnapshot(addRows).build();
+
+    assertThat(firstAddition.nextRowId()).isEqualTo(newRows);
+
+    Snapshot addMoreRows =
+        new BaseSnapshot(
+            1, 2, 1L, 0, DataOperations.APPEND, null, 1, "foo", firstAddition.nextRowId(), newRows);
+
+    TableMetadata secondAddition =
+        TableMetadata.buildFrom(firstAddition).addSnapshot(addMoreRows).build();
+
+    assertThat(secondAddition.nextRowId()).isEqualTo(newRows * 2);
+  }
+
+  @TestTemplate
+  public void testInvalidSnapshotAddition() {
+    assumeThat(formatVersion).isGreaterThanOrEqualTo(TableMetadata.MIN_FORMAT_VERSION_ROW_LINEAGE);
+
+    Long newRows = 30L;
+
+    TableMetadata base = baseMetadata();
+
+    Snapshot invalidLastRow =
+        new BaseSnapshot(
+            0, 1, null, 0, DataOperations.APPEND, null, 1, "foo", base.nextRowId() - 3, newRows);
+
+    assertThatThrownBy(() -> TableMetadata.buildFrom(base).addSnapshot(invalidLastRow))
+        .isInstanceOf(ValidationException.class)
+        .hasMessageContaining("Cannot add a snapshot whose 'first-row-id'");
+
+    Snapshot invalidNewRows =
+        new BaseSnapshot(
+            0, 1, null, 0, DataOperations.APPEND, null, 1, "foo", base.nextRowId(), null);
+
+    assertThatThrownBy(() -> TableMetadata.buildFrom(base).addSnapshot(invalidNewRows))
+        .isInstanceOf(ValidationException.class)
+        .hasMessageContaining(
+            "Cannot add a snapshot with a null 'added-rows' field when row lineage is enabled");
+  }
+
+  @TestTemplate
+  public void testFastAppend() {
+    assumeThat(formatVersion).isGreaterThanOrEqualTo(TableMetadata.MIN_FORMAT_VERSION_ROW_LINEAGE);
+
+    TestTables.TestTable table =
+        TestTables.create(
+            tableDir, "test", TEST_SCHEMA, PartitionSpec.unpartitioned(), formatVersion);
+    TableMetadata base = table.ops().current();
+    table.ops().commit(base, TableMetadata.buildFrom(base).enableRowLineage().build());
+
+    assertThat(table.ops().current().rowLineageEnabled()).isTrue();
+    assertThat(table.ops().current().nextRowId()).isEqualTo(0L);
+
+    table.newFastAppend().appendFile(fileWithRows(30)).commit();
+
+    assertThat(table.ops().current().rowLineageEnabled()).isTrue();
```

Comment:
My general 2c, not sure I understand the intent behind the assertions in this test class that row lineage is still enabled after the commit. I feel like for the purpose of this test class for the happy cases is just to make sure the first row/next row IDs are set properly according to the operation (that implicitly verifies that row lineage is enabled).

Reply:
Yeah, i'm just being paranoid. Making sure it isn't turned off or anything, I can remove that check